vidyut-lipi/src/bin/transliterate.rs

```diff
@@ -1,35 +1,453 @@
+// translit.rs
+//
+
+use vidyut_lipi as pj;
 use clap::Parser;
-use std::error::Error;
-use vidyut_lipi::transliterate;
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::path::Path;
+use std::io::Write;
+
 
-#[derive(Parser)]
-#[command(author, version, about)]
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
 struct Args {
-    /// Text to transliterate
-    #[arg(long)]
-    text: String,
+    #[arg(short, long)]
+    mode: String,
+    #[arg(short, long)]
+    infile: Option<std::path::PathBuf>,
+    #[arg(short, long)]
+    outfile: Option<std::path::PathBuf>,
+    #[arg(short, long)]
+    frommap: String,
+    #[arg(short, long)]
+    tomap: String,
+    #[arg(short, long)]
+    data: Option<String>,
 }
 
-fn run(args: Args) -> Result<(), Box<dyn Error>> {
-    let result = transliterate(
-        &args.text,
-        vidyut_lipi::Scheme::Iast,
-        vidyut_lipi::Scheme::Slp1,
-    );
+fn main() {
 
-    println!("{}", result);
+    let mut from_map: pj::TranslationMap =  pj::TranslationMap::default();
+    let mut to_map: pj::TranslationMap = pj::TranslationMap::default();
 
-    Ok(())
+    let args = Args::parse();
+
+    match args.frommap.as_str() {
+	"itrans" => {
+	    pj::generate_map("itrans.json", &mut from_map);
+	},
+	"slp1" => {
+	    pj::generate_map("slp1.json", &mut from_map);
+	}
+	_ => {
+	    println!("Valid from maps are itrans and slp1");
+	    return;
+	}
+    }
+    
+    match args.tomap.as_str() {
+	"devanagari" => {
+	    pj::generate_map("devanagari.json", &mut to_map);
+	},
+	"malayalam" => {
+	    pj::generate_map("malayalam.json", &mut to_map);
+	}
+	_ => {
+	    println!("Valid from maps are devanagari and malayalam");
+	    return;
+	}
+    }
+
+    match args.mode.as_str() {
+	"file" => {
+	    if args.infile.is_some() && args.outfile.is_some() {
+		transliterate_file(args.infile.unwrap(), args.outfile.unwrap(),
+				   from_map, to_map);	    
+	    }
+	},
+	"text" => {
+	    let mut outstr:String = "".to_string();
+	    if args.data.is_some() {
+		pj::transliterate(args.data.unwrap(), &from_map, &to_map, &mut outstr);
+		println!("{}", outstr);		
+	    } else {
+		println!("No data available...");
+	    }
+	},
+	_ => {
+	    println!("valid modes are file / text");
+	}
+    }
+
+
+    
 }
 
-fn main() {
-    let args = Args::parse();
+// The output is wrapped in a Result to allow matching on errors
+// Returns an Iterator to the Reader of the lines of the file.
+fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
+where P: AsRef<Path>, {
+    let file = File::open(filename)?;
+    Ok(io::BufReader::new(file).lines())
+}
+
+// Right now assume we process only itx files.
+// handler functions might b required for individual file types. 
+fn transliterate_file<P>(inf: P, outf:P, inmap:pj::TranslationMap, outmap:pj::TranslationMap)
```

Comment:
Are itx files widely used? I recall seeing them only on sanskritdocuments.org. If the file format is rare, I think we should keep this code fairly general and just assume plain text.

Reply:
Well the document that i had was ITX. So my immediate need was to go from itrans->devanagari. i would imagine there would be individual handlers for different file types - mainly to strip out any of the markup tags.

Reply:
I see -- I've been thinking of this binary as a testing tool. Are you thinking of it as a standalone tool?

I think in practice most people will use a web transliterator like Sanscript to transliterate a document -- it's fast enough and easier for non-technical users.